File: 144.BinaryTreePreorderTraversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        
        List<Integer> res = new ArrayList<Integer>();
        if(root == null){
            return res;
        }        
        res.add(root.val);
        List<Integer> left = preorderTraversal(root.left);
        List<Integer> right = preorderTraversal(root.right);
        
        if (left != null){
            res.addAll(left);            
        }
        if (right != null){
            res.addAll(right);            
        }    
        
        return res;
        
    }
}

File: 144.BinaryTreePreorderTraversal/Solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

//recursion version
func preorderTraversal(root *TreeNode) []int {
    if root == nil{return nil}

    res := []int{root.Val}
    l := preorderTraversal(root.Left)
    r := preorderTraversal(root.Right)
    
    res = append(res, l...)
    res = append(res, r...)
    
    return res
}
